Fixed #27910 -- Added enumeration helpers for use in Field.choices

pope1ni opened the pull request:
Ticket [#27910](https://code.djangoproject.com/ticket/27910)
Supersedes PR [#11223](https://github.com/django/django/pull/11223)

Improvements over the previous pull request:
- Simplified definition by moving everything from `ChoiceEnum` into `ChoiceEnumMeta`.
- Added additional properties for fetching `labels`, `values`, and `names`.
- Used [`enum.unique()`](https://docs.python.org/3/library/enum.html#enum.unique) to prevent duplicate values.
- Moved all documentation from `django.db.models.enums` into `docs/ref/models/fields.txt`.
- Improved and expanded documentation in `docs/ref/models/fields.txt`.
- Improved and expanded tests in `tests/model_fields/test_choiceenum.py`.
- Improved messages in exceptions.

Things left to do:
- [x] Update documentation in `docs/topics/db/models.txt`.
- [ ] ~~Add documentation for the `.validate()` method.~~
- [x] Check whether the [Functional API](https://docs.python.org/3/library/enum.html#functional-api) works; add tests and documentation.
- [x] Check handling of blank values in choices, e.g. `''` or `None`.
  - _Handling `None` is potentially problematic as were restricted to `int` or `str`._

pope1ni commented:
@shaib: Could you explain the use case for the `.validate()` method?

@felixxm: This is largely done. The main issue remaining is handling of blank values using `None` which doesn't work because `ChoiceIntEnum` subclasses `int` and `ChoiceStrEnum` subclasses `str`. For `int`, using `None` blows up, and for `str`, using `None` coerces it to `'None'`. I've added some expected failure test cases. We have the following options:
1. See if we can define some sort of composite `NullableInt` and `NullableStr` types.
2. Use some sentinel values that are valid that can be forced to be `None` somehow.
3. Document that overriding the empty value to use a custom label is not supported.

It would be a shame if #3 were the only option here.

felixxm commented on django/db/models/enums.py:
```diff
@@ -0,0 +1,64 @@
+                try:
+                    value, label = value
+                except ValueError as e:
+                    raise e.__class__(f'Invalid member for {classname}: {key}.') from e
```
Please don't use f-string, for now we don't allow them in Django's codebase.

Is there any reason to use `e.__class__` instead of `ValueError`?

pope1ni replied:
> Please don't use f-string, for now we don't allow them in Django's codebase.

Sorry. I misunderstood https://github.com/django/django/pull/11223#discussion_r279075913.

> Is there any reason to use `e.__class__` instead of `ValueError`?

I guess not in this case as we're only catching one type of exception. Sometimes it is useful if you want to keep the same exception class but change the message as in this case. Will change.

felixxm commented on django/db/models/enums.py:
```diff
@@ -49,7 +49,7 @@ def validate(cls, value):
             cls(value)
         except ValueError:
-            raise exceptions.ValidationError(_(f'Invalid choice: {cls.__name__}({value!r})'))
+            raise exceptions.ValidationError(_('Invalid choice: %s(%r)' % (cls.__name__, value)))
```
`%s(%r)` -> `%s (%r)`

pope1ni replied:
Hmm, could do. I think that it was `%s(%r)` because `Fruit('APPLE')` is equivalent to `Fruit.APPLE`.

Otherwise it could have been rewritten as `'Invalid choice for %s: %r.'`

felixxm replied:
Ahh OK, so maybe `%s.%s`.

pope1ni replied:
Sorry - I confused here. It is trying to show that passing an invalid member name to the enum is what has caused the validation error. But I guess we are trying to hide that as an implementation detail, so we could go for my alternative.

I'm going to investigate using ``__contains__``, however, as this was an attempt to provide an alternate way to work around ``'APPLE' in Fruit``.

shaib commented:
> @shaib: Could you explain the use case for the `.validate()` method?
> 

It was motivated by https://groups.google.com/d/msgid/django-developers/CAL13Cg_KEtgmNzRo%3DC8cnCiNsukr5YOCTv3MF8AybV%3D%3DiUXP0g%40mail.gmail.com

> @felixxm: This is largely done. The main issue remaining is handling of blank values using `None` which doesn't work because `ChoiceIntEnum` subclasses `int` and `ChoiceStrEnum` subclasses `str`. For `int`, using `None` blows up, and for `str`, using `None` coerces it to `'None'`. I've added some expected failure test cases. We have the following options:
> 
>     1. See if we can define some sort of composite `NullableInt` and `NullableStr` types.
> 
>     2. Use some sentinel values that are valid that can be forced to be `None` somehow.
> 

No, there's no need for these. Take a look at the loop starting at [line 14 of enums.py](https://github.com/django/django/pull/11540/files#diff-18d0038e6938e1f90be081169dae9db2R14) -- it clearly shows that the values assigned to enum members in the source don't need to be valid members of the type (tuples are not `int`s). We can decide that e.g. if we reach line 26 with `value is None` then, instead of `dict.__setitem__(classdict, key, value)` we do `dict.__delitem__(classdict, key)`. The only difference is that the properties will need to use `cls._value2label_map_.keys()` instead of `cls.__members__.values()`.

I think an even better variation of this idea is, instead of using the `None` value as sentinel, to use a special member name -- say, `__empty__` or `__no_choice__`; this way, the user doesn't need to repeat `None`,  doesn't have to invent a name that will never be used, and gets the right clue that this is not going to be a member of the enum.

pope1ni commented:
I've pushed a `fixup!` which removes `validate()` and implements `__contains__()` to make things work as expected according to James' [comment](https://groups.google.com/forum/#!msg/django-developers/bCJ28uz0QFE/PsLYzM-hEQAJ):

```python
>>> class Suit(models.ChoiceIntEnum):
...     DIAMOND = 1, _('Diamond')
...     SPADE = 2, _('Spade')
...     HEART = 3, _('Heart')
...     CLUB = 4, _('Club')
...
>>> Suit.DIAMOND in Suit
True
>>> 1 in Suit
True
>>> 0 in Suit
False
```

I will look into the label for the empty value soon.

pope1ni commented:
@felixxm @shaib I think that last major things have been addressed here and we're ready for review.

pope1ni commented on django/db/models/enums.py:
```diff
@@ -0,0 +1,65 @@
+            dict.__setitem__(classdict, key, value)
+        cls = super().__new__(metacls, classname, bases, classdict)
+        cls._value2label_map_ = choices
+        cls.label = property(cls._value2label_map_.get)
```
> It's very elegant…

I thought so!

> …, but a little obscure, and deserves a comment to explain it…

That's fair. I'll try to get round to that today.

@felixxm Any chance that you could look at this one too? Thanks 😊

pope1ni replied:
Comment added.

shaib replied:
:+1:

felixxm replied:
Looks good.

carltongibson commented on docs/ref/models/fields.txt:
```diff
@@ -94,38 +94,86 @@ and the second element is the human-readable name. For example::
-Generally, it's best to define choices inside a model class, and to
-define a suitably-named constant for each value::
+Django provides special enumeration types you can subclass to define this
+iterable in a concise way::
```
> Generally, it's best to define choices inside a model class, and to
define a suitably-named constant for each value

I think this particular point has been lost, and is important. 

Later you `Card.Suit` being declared as a nested class, but it's only implicit what's going on. 

I **think** I'd leave the `YEAR_IN_SCHOOL_CHOICES` examples as are, through to the `self.JUNIOR` usage example, and then introduce Enum approach. (But not hung up on that particular ordering...)

pope1ni replied:
Thanks for the review. I restored this section and then introduced the enum approach as suggested.

carltongibson commented on docs/ref/models/fields.txt:
```diff
@@ -119,14 +122,91 @@ define a suitably-named constant for each value::
+
+    from django.utils.translation import gettext_lazy as _
+
+    class YearInSchool(models.ChoiceStrEnum):
```
So... do we want `YearInSchool` as a nested class inside `Student`, since we kind of recommend that usage?

pope1ni replied:
We can do. Sometimes people want to reuse the same set of choices for multiple models. Feel free to move it under `Student` if you think it makes the examples clearer.

pope1ni replied:
Ok. Have made the change.

pope1ni commented:
Have rebased this again. Is there anything else to do @carltongibson / @felixxm?

shaib commented:
@frnhr I'm getting the same error when trying to define a date-based enum with no relation to ChoiceEnum:
```pycon
>>> from enum import Enum
>>> from datetime import date
>>> class MyEnum(date, Enum):
...   A = date(2014, 1, 1)
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 218, in __new__
    enum_member = __new__(enum_class, *args)
TypeError: an integer is required (got type datetime.date)
```
So I'm guessing there's little we can do here. I would try something like using the `str()` of the dates as values, but that would need to be taken care of in `ChoiceDateEnum`'s metaclass.

In my taste, this use-case falls a little out of what Django should actively support.

frnhr commented:
Fair enough, but we should at least mention this in the docs. I doubt that I will be the only one attempting to use it with a field that is not `int` or `str`.

Especially because it _looks_ like it should work, and it is documented right next to the old-style choices which _do work_ for all cases.

pope1ni commented:
Thanks @shaib for posting the info - you got there ahead of me. I'd noticed the same issue.

I had a look if this were a problem due to some order of base classes, but there is a strict order these must be passed in so we are doing things the right way:
```python
>>> import datetime
>>> import enum
>>> class X(enum.Enum, datetime.date):
...     pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 130, in __prepare__
    member_type, first_enum = metacls._get_mixins_(bases)
  File "/usr/lib/python3.7/enum.py", line 467, in _get_mixins_
    raise TypeError("new enumerations should be created as "
TypeError: new enumerations should be created as `EnumName([mixin_type, ...] [data_type,] enum_type)`
```

I checked with a bunch of other common datatypes that might be used with Django.

The following worked when creating a simple enum:
```python
>>> import decimal
>>> import enum
>>> import ipaddress
>>> class X(bytes, enum.Enum):
...     A = b'raw'
... 
>>> class X(float, enum.Enum):
...     A = 1.5
... 
>>> class X(dict, enum.Enum):
...     A = {'A': 1}
... 
>>> class X(list, enum.Enum):
...     A = [1, 2, 3]
... 
>>> class X(set, enum.Enum):
...     A = {1, 2, 3}
... 
>>> class X(tuple, enum.Enum):
...     A = (1, 2, 3)
... 
>>> class X(decimal.Decimal, enum.Enum):
...     A = decimal.Decimal('1.0')
...
>>> class X(ipaddress.IPv4Address, enum.Enum):
...     A = ipaddress.IPv4Address('127.0.0.1')
... 
>>> class X(ipaddress.IPv6Address, enum.Enum):
...     A = ipaddress.IPv6Address('::1')
... 
```

The following didn't work for varying reasons:
```python
>>> import datetime
>>> import enum
>>> import uuid
>>> class X(bool, enum.Enum):
...     pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 170, in __new__
    enum_class = super().__new__(metacls, cls, bases, classdict)
TypeError: type 'bool' is not an acceptable base type
>>> class X(datetime.date, enum.Enum):
...     A = datetime.date.today()
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 218, in __new__
    enum_member = __new__(enum_class, *args)
TypeError: an integer is required (got type datetime.date) 
>>> class X(datetime.datetime, enum.Enum):
...     A = datetime.date.now()
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 218, in __new__
    enum_member = __new__(enum_class, *args)
TypeError: an integer is required (got type datetime.datetime)
>>> class X(datetime.time, enum.Enum):
...     A = datetime.time(1, 2, 3)
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 218, in __new__
    enum_member = __new__(enum_class, *args)
TypeError: an integer is required (got type datetime.time)
>>> class X(datetime.timedelta, enum.Enum):
...     A = datetime.timedelta(1)
...  
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 218, in __new__
    enum_member = __new__(enum_class, *args)
TypeError: unsupported type for timedelta days component: datetime.timedelta
>>> class X(datetime.timezone, enum.Enum):
...     A = datetime.timezone.utc
...  
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 170, in __new__
    enum_class = super().__new__(metacls, cls, bases, classdict)
TypeError: type 'datetime.timezone' is not an acceptable base type
>>> class X(uuid.UUID, enum.Enum):
...     A = uuid.uuid4()
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 216, in __new__
    enum_member._value_ = value
  File "/usr/lib/python3.7/uuid.py", line 262, in __setattr__
    raise TypeError('UUID objects are immutable')
TypeError: UUID objects are immutable
```

I don't know whether this is a bug in the enum implementation - perhaps a bug should be opened to get this addressed in Python itself. Note that, in theory, we don't need to subclass the data type, but we've done that to try and make sure that things behave as expected when comparing enum members to values of the data type. With that caveat, it is possible to do the following:

```python
>>> import datetime
>>> import enum
>>> from django.db.models.enums import ChoiceEnumMeta
>>> class Date(enum.Enum, metaclass=ChoiceEnumMeta):
...     TODAY = datetime.date.today()
... 
>>> Date.TODAY == datetime.date.today()  # Won't work...
False
>>> Date.TODAY.value == datetime.date.today()  # Still works...
True
```

In light of this, I have made one change that ensures that `.label` will work when the data type is not provided.

I think providing `int` and `str` will cover >95% of use cases, and using an enum is not a requirement.

pope1ni commented:
@shaib, @frnhr So I was reading into this a little more, and realised that our approach was wrong after seeing [this example](https://docs.python.org/3.7/library/enum.html#timeperiod). We shouldn't be using `datetime.date(...)` as the value of each member, but a tuple of the arguments to it, e.g.

```python
class MoonLandings(datetime.date, enum.Enum, metaclass=ChoiceEnumMeta):
    APOLLO_11 = 1969, 7, 20, _('Apollo 11 (Eagle)')
    APOLLO_12 = 1969, 11, 19, _('Apollo 12 (Intrepid)')
    APOLLO_14 = 1971, 2, 5, _('Apollo 14 (Antares)')
    APOLLO_15 = 1971, 7, 30, _('Apollo 15 (Falcon)')
    APOLLO_16 = 1972, 4, 21, _('Apollo 16 (Orion)')
    APOLLO_17 = 1972, 12, 11, _('Apollo 17 (Challenger)')
```

So the challenge is making this work properly with the extra label. Am going to work this out and add a bunch of tests for various different types.

pope1ni commented:
Well that was all manner of magic and weird when looking through the internals of `Enum`. The examples I added using classes from the `ipaddress` module behaved differently which broke our implementation of `.label` which I have fixed that properly now.

So, more tests for custom subclasses, various fixes, etc. and we should be good for many more custom implementations :tada: 

Thanks @frnhr for the challenge!

I still want to add a few more tests for enums of `list`, `set`, `dict`, `tuple`, and `datetime.datetime`. (It may also be interesting to see how things work out with `enum.IntFlag`, but that shouldn't be considered a blocker for this.)

Any further reviews are welcome - I'd like to get this in for 3.0

pope1ni commented:
Ok. So I have also added tests for `datetime.datetime` and `frozenset`.

Currently `dict`, `list`, and `set` are a bit awkward as they are unhashable we are using a dict for looking up labels from values. It also looks as though `tuple` will need some special handling because we are assuming that the last item in the tuple is the label when it is a (lazy) string. In the tuple case this is very ambiguous.

I don't think this is a deal breaker as I can't imagine that these would be used very often as choice values, and besides, it is possible to fall back to using lists of 2-tuples.

felixxm commented on tests/model_fields/test_choiceenum.py:
```diff
@@ -0,0 +1,282 @@
+Place = models.ChoiceIntEnum('Place', 'FIRST SECOND THIRD')
+
+
+class Separator(bytes, enum.Enum, metaclass=ChoiceEnumMeta):
```
We added many tests with custom enums that use `metaclass=ChoiceEnumMeta` if we want to support this usage then documentation is missing (at least a short section with a single example, maybe `MoonLandings`).

pope1ni replied:
`ChoiceIntEnum` and `ChoiceStrEnum` were explicitly defined to avoid the more complex signature, but they are not compulsory. Enums can be subclassed unless they have members defined in them. We could have skipped defining them (although for `str` we wanted to define `_generate_next_value_()` to make the Functional API work sanely out of the box).

Of the various tested examples, are there any data types that we want to provide an implementation for like `ChoiceIntEnum` and `ChoiceStrEnum`? Also do we like those names, or would we prefer `IntegerChoiceEnum` and `TextChoiceEnum` or even just `IntegerChoices` and `CharChoices` / `TextChoices` as this would tie up nicely with the fields they are intended for?

shaib replied:
I don't think we should include other choice enums out of the box -- but maybe that's just me. On the other side, if possible, I'd rather people not have to use `ChoiceEnumMeta` explicitly -- I prefer 
```python
class Separator(bytes, ChoiceEnum):
```
if it can be arranged.

With respect to the names, I took the cue from `enum.IntEnum`, thinking that this is a cute hack on enums rather than a major feature for fields. If we see it as an important Django feature, perhaps the naming you suggest is preferable. No strong opinion either way.

pope1ni replied:
Ok. So I agree that we want to make this as simple for people as possible, and to that end after some more thought I think that the names should also change. After all, these custom enum implementations are designed to be used for the specific purpose of working with `.choices`.

```python
class ChoicesMeta(enum.EnumMeta):
    ...


class Choices(enum.Enum, metaclass=ChoicesMeta):
    pass


class IntegerChoices(int, Choices):
    ...


class TextChoices(str, Choices):
    ...
```
So usage would then be quite descriptive:
```python
class Deck(models.Model):
    class Suit(models.IntegerChoices):
        DIAMOND = 1
        SPADE = 2
        HEART = 3
        CLUB = 4

    suit = models.PositiveIntegerField(choices=Suit.choices)
    ...


class Event(models.Model):
    Medal = models.TextChoices('Medal', 'GOLD SILVER BRONZE')

    name = models.CharField(max_length=100)
    date = models.DateField()
    medal = models.CharField(choices=Medal.choices)
    ...
```

And this would also make defining new choices very simple:
```python
class Separator(bytes, models.Choices):
    ...
```

shaib replied:
Works for me.

felixxm commented on tests/model_fields/test_choiceenum.py:
```diff
@@ -0,0 +1,282 @@
+
+
+class YearInSchool(models.ChoiceStrEnum):
+    FRESHMAN = 'FR', _('Freshman')
```
I don't think that we need to use `gettext_lazy` in all enums.

pope1ni replied:
Will leave for `Suit` and `YearInSchool` as this is explicitly tested for each of `ChoiceIntEnum` and `ChoiceStrEnum`, but will remove all the others.

pope1ni replied:
Will leave for `Suit` and `YearInSchool` as this is explicitly tested for each of `ChoiceIntEnum` and `ChoiceStrEnum`, but will remove all the others.

felixxm commented on django/db/models/enums.py:
```diff
@@ -0,0 +1,76 @@
+    @property
+    def choices(cls):
+        empty = [(None, cls.__empty__)] if hasattr(cls, '__empty__') else []
+        return empty + [(member.value, member.label) for member in cls.__members__.values()]
```
I think we can use `in cls` instead of `in cls.__members__.values()` (here and in `names`).

felixxm commented on django/db/models/enums.py:
```diff
@@ -0,0 +1,76 @@
+    @property
+    def values(cls):
+        empty = [None] if hasattr(cls, '__empty__') else []
+        return empty + [member.value for member in cls.__members__.values()]
```
I would use `choices` in `labels` and `values` to simplify implementation, e.g.
```python
    @property
    def labels(cls):
        return [label for _, label in cls.choices]
        
    @property
    def values(cls):
        return [value for value, _ in cls.choices]
```

shaib commented:
I know that it was I who suggested the name `__empty__`, but on second thought, I worry that this is a name that might be used by Python itself in the future. Maybe it would be better to use something like `_empty_` or `__empty`.

On the other hand, we do seem to have `__html__`... So, I don't know.

felixxm commented:
We also use `__all__` :smile:  so I'm fine with `__empty__`.

shaib commented:
`__all__` is used in Django for a Django-specific meaning?

It is used by Python, and has been forever.

felixxm commented:
@shaib As a string constant in forms `ALL_FIELDS = '__all__'` (I know it's not the same). After reconsideration I agree that we could use the different name e.g. `_empty_`.

pope1ni commented:
So it seems that enums also define some special `_sunder_` names as well as the `__dunder__` names that are common, and given the number of them and the sorts of names they have - [see here](https://docs.python.org/3/library/enum.html#finer-points) - I feel that `_empty_` is at more risk of collision than `__empty__`. In fact, we're not allowed to define our own `_sunder_` names:

```python
>>> import enum
>>> class X(enum.Enum):
...     _empty_ = '(Empty)'
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 2, in X
  File "/usr/lib/python3.7/enum.py", line 84, in __setitem__
    raise ValueError('_names_ are reserved for future Enum use')
ValueError: _names_ are reserved for future Enum use
```

I tried a few alternatives:

```python
>>> import enum
>>> class X(enum.Enum):
...     _empty = '(Empty)'
...     __empty = '(Empty)'
...     __empty__ = '(Empty)'
... 
>>> type(X._empty)
<enum 'X'>
>>> type(X.__empty)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 349, in __getattr__
    raise AttributeError(name) from None
AttributeError: __empty
>>> type(X.__empty__)
<class 'str'>
```

So we can't use `_empty` or `__empty`. We could in theory use [`_ignore_`](https://docs.python.org/3/library/enum.html#supported-sunder-names) to prevent something becoming a member, but that is 3.7+ only and also removes the attribute from the final class. So we're stuck with using `__empty__` I think...

shaib commented:
> ```python
> >>> import enum
> >>> class X(enum.Enum):
> ...     __empty = '(Empty)'
> >>> type(X.__empty)
> [...]
> AttributeError: __empty

That's just the Python prefix-`__`-is-mangled-to-fake-private mechanism:
```pycon
>>> import enum
>>> class X(enum.Enum):
...   __empty = 'Nada'
... 
>>> X.__empty
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.7/enum.py", line 349, in __getattr__
    raise AttributeError(name) from None
AttributeError: __empty
>>> X._X__empty
<X._X__empty: 'Nada'>
```
But I agree, something like this should not be used.

pope1ni commented:
Ah, yes. Indeed.

pope1ni commented:
Ok. I'll look to add the additional documentation and tweak the release notes a bit later.

pope1ni commented:
@shaib Do you have any thoughts on the second paragraph of https://github.com/django/django/pull/11540#discussion_r320260550?

pope1ni commented:
Also, what do we think about a `.max_length` property (at least for the `ChoiceStrEnum`)?

```python
class Event(models.Model):
    Medal = models.ChoiceStrEnum('Medal', 'GOLD SILVER BRONZE')

    name = models.CharField(max_length=100)
    date = models.DateField()
    medal = models.CharField(choices=Medal.choices, max_length=Medal.max_length)
```

If I had a penny for each time I've seen `choices` change and `max_length` subsequently being too small... The main challenge would be the removal of the longest string where that value is still in the table - but in that case the migration should fail. Obviously use of this wouldn't be compulsory, but I can see it being *really* helpful.

shaib commented:
Setting the `max_length` of the field according to choices seems wrong to me. It's fine to do on PG where max_length is only a validation, but on other engines where it actually dictates layout, changing max_length can cause a very painful migration. Additionally, choices are not really enforced, so it is likely that values other than the choices are stored in the field (the way you mentioned, with a removed choice, or otherwise), and when that happens, it is much more likely to happen in production than in dev/test environments.

There should be a check that all choices actually fit in the field, but I don't think Django should promote a practice of letting the choices set the field length.

pope1ni commented:
> It's fine to do on PG where max_length is only a validation...

```sql
postgres=# create table t (c varchar(5));
CREATE TABLE
Time: 369.761 ms
postgres=# insert into t (c) values ('abcdefghij');
ERROR:  value too long for type character varying(5)
Time: 0.410 ms
postgres=# insert into t (c) values ('abcde');
INSERT 0 1
Time: 62.839 ms
postgres=# alter table t alter column c type varchar(3);
ERROR:  value too long for type character varying(3)
Time: 73.806 ms
postgres=# alter table t alter column c type varchar(3) USING c :: varchar(3);
ALTER TABLE
Time: 120.832 ms
postgres=# select * from t;
  c  
─────
 abc
(1 row)

Time: 0.489 ms
```
Performance-wise in PostgreSQL there is actually no difference between the [different text types](https://wiki.postgresql.org/wiki/FAQ#What_is_the_difference_between_the_various_character_types.3F). I don't believe Django allows not setting `max_length`, otherwise I probably wouldn't in a lot of cases. Changing the `max_length` will still cause a migration - am not sure whether PostgreSQL has to rewrite parts of the table or not if just expanding.

> There should be a check that all choices actually fit in the field, but I don't think Django should promote a practice of letting the choices set the field length.

OK, agreed. Let's leave it to a system check. Will look to try and add one later in a separate PR because this is always a source of pain.

shaib commented:
Just to clarify my PG comment: This was all about the database engines.  The fact that Django forces you to specify a `max_length` on a `CharField` is unrelated.

On Postgres, the internal representation of different text types is the same (that's why the performance is the same). So when you change from `varchar(5)` to `varchar(10)` all that changes is the validation rules. This is an easy migration on the database side, the only change is in metadata. There can be validations and castings, which can be lengthy, but there can be all kinds of mitigations and the whole thing is generally not very painful.

Oracle, for example (at least in the versions of 6-7 years ago when I was using it) was actually keeping rows together, and allocating the right number of characters in each row for each field. Thus, a similar change implies a change in the layout of the table's storage, which can be very expensive -- and could mean a long downtime for the migration.

SQLite isn't usually discussed much in terms of effects on production systems, but just for good measure -- last I checked, every such change implied a complete recreation of the table.

pope1ni commented:
Ok. Changes made. Thanks Shai.

pope1ni commented:
> OK, agreed. Let's leave it to a system check. Will look to try and add one later in a separate PR because this is always a source of pain.

Created PR #11742.

felixxm commented:
@pope1ni I pushed minor edits, rebased from master, simplified tests, and moved choices tests to a separate directory.

I noticed that [these lines](https://github.com/django/django/pull/11540/files#diff-18d0038e6938e1f90be081169dae9db2R22-R23) are not covered, I don't think that we still need this catch or if we really want to catch `ValueError` then we should do this in
```python
super().__new__(metacls, classname, bases, classdict)
```

I'm working on docs.

pope1ni commented on tests/model_enums/tests.py:
```diff
@@ -0,0 +1,249 @@
+    LINK_LOCAL_UNICAST = 'fe80::/10', 'Link-Local Unicast'
+
+
+class TestCustomChoiceEnum(SimpleTestCase):
```
`TestCustomChoiceEnum` → `CustomChoicesTests`

pope1ni commented on docs/releases/3.0.txt:
```diff
@@ -81,6 +81,42 @@ Expressions that outputs :class:`~django.db.models.BooleanField` may now be
 used directly in ``QuerySet`` filters, without having to first annotate and
 then filter against the annotation.
 
+Enumerations for model choices
```
`model` → `model field`

pope1ni commented:
> I noticed that [these lines](https://github.com/django/django/pull/11540/files#diff-18d0038e6938e1f90be081169dae9db2R22-R23) are not covered, I don't think that we still need this catch...

So originally when we only covered `int` and `str`, we restricted this to expect `value, label` and there were some tests for too few or too many items. Because enum values are not an instance of the concrete data type, but a tuple of arguments passed to the constructor of that concrete data type, we changed tack. As such, yes, these lines are obsolete - we rely on the underlying data type to blow up if the arguments are invalid. We're just assuming that the final item in the tuple for the member value is going to be the label if it is a (lazy) string.

felixxm commented:
@pope1ni OK, we are almost there, I pushed all edits, but I also noticed one issue with migrations that I would try to fix, i.e. `default` is not evaluated and migrations crashes, e.g.
```python
    field=models.CharField(choices=[('FR', 'Freshman'), ('SO', 'Sophomore'), ('JR', 'Junior'), ('SR', 'Senior'), ('GR', 'Graduate'), ('XX', 'Xxx')], default=test_one.models.YearInSchool('FR'), max_length=2),
AttributeError: module 'test_one.models' has no attribute 'YearInSchool'
```

the old-fashioned choices evaluates properly to the `default='FR'`.

shaib commented:
That's not an issue with default per se, it's an issue with defining the choices enum within the model class, and the auto-migration-composer not handling nested classes.

pope1ni commented:
I think that could be handled by overriding `__repr__` to use `__qualname__`? This will also work if default is defined to be `YearInSchool.FRESHMAN.value`.

pope1ni commented on docs/ref/models/fields.txt:
```diff
@@ -119,14 +122,107 @@ define a suitably-named constant for each value::
+
+.. class:: Choices(**options)
+.. class:: TextChoices(**options)
+.. class:: IntegerChoices(**options)
```
Does having these here break the flow of the section?

felixxm replied:
Yes a bit, I'm going to remove them.

felixxm replied:
We can decide to add `docs/ref/enums` in the future, but IMO it's fine to keep this inline for now.

pope1ni replied:
My thoughts exactly. Thanks Felix.

felixxm commented:
@shaib Yes, I was able to fix this quite easily with:
```diff
diff --git a/django/db/migrations/serializer.py b/django/db/migrations/serializer.py
index 1f1b3f4f20..c6bfd90621 100644
--- a/django/db/migrations/serializer.py
+++ b/django/db/migrations/serializer.py
@@ -45,6 +45,9 @@ class BaseSimpleSerializer(BaseSerializer):
     def serialize(self):
         return repr(self.value), set()
 
+class ChoicesSerializer(BaseSerializer):
+    def serialize(self):
+        return repr(self.value.value), set()
 
 class DateTimeSerializer(BaseSerializer):
     """For datetime.*, except datetime.datetime."""
@@ -279,6 +282,7 @@ class Serializer:
         set: SetSerializer,
         tuple: TupleSerializer,
         dict: DictionarySerializer,
+        models.Choices: ChoicesSerializer,
         enum.Enum: EnumSerializer,
         datetime.datetime: DatetimeDatetimeSerializer,
         (datetime.date, datetime.timedelta, datetime.time): DateTimeSerializer,
```
after that it evaluates to `default='FR'`. What do you think?

pope1ni commented:
> What do you think?

Good idea. Should seamlessly work for custom choices then too. Can you try with one that subclasses `datetime.date` or `ipaddress.IPv4Address` to see if imports are resolved appropriately?

felixxm commented:
Yes, I'm preparing tests.

felixxm commented:
I pushed fix.

felixxm commented:
Thanks y'all :rocket: Many thanks to Nick and Shai :bowing_man: 

Now we can celebrate :champagne: :dancer: :rocket: :confetti_ball:

shaib commented:
Many thanks, @pope1ni and @felixxm! When I started this, I never thought it would make it to the list of major features of the release... but you have grown it bigger than I expected. It is a reason for celebration indeed.

pope1ni commented:
Thanks both of you!
